### PR TITLE
Fix errorbar property cycling to match plot.

### LINKED ISSUE #7074: Using a linestyle cycler with plt.errorbar results in strange plots
If the linestyles used in a `plt.errorbar` plot are adjusted manually, results are as expected. However, when using a `linestyle` cycler, the linestyles are simply ignored whereas using a `ls` cycler results in errorbar end points being connected by lines.

Consider the following minimal example

``` python
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
from cycler import cycler

x = np.arange(0.1, 4, 0.5)
y = np.exp(-x)
offsets = [0, 1]


linestyles = ['-', '--']

plt.figure()
for offset, linestyle in zip(offsets, linestyles):
    plt.errorbar(x, y + offset, xerr=0.1, yerr=0.3, ls=linestyle, color='b')
plt.title('manual plots')
plt.savefig('manual_plots.png')


linestyle_cycle = cycler('linestyle', linestyles)
mpl.rcParams['axes.prop_cycle'] = linestyle_cycle

plt.figure()
for offset in offsets:
    plt.errorbar(x, y + offset, xerr=0.1, yerr=0.3, color='b')
plt.title('Linestyle cycler plots')
plt.savefig('linestyle_cycler_plots.png')


ls_cycle = cycler('ls', linestyles)
mpl.rcParams['axes.prop_cycle'] = ls_cycle

plt.figure()
for offset in offsets:
    plt.errorbar(x, y + offset, xerr=0.1, yerr=0.3, color='b')
plt.title('ls cycler plots')
plt.savefig('ls_cycler_plots.png')
```

which produces the following set of figures

manual_plots
linestyle_cycler_plots
ls_cycler_plots

I ran the above example using Anaconda 4.1.1 with Python 3.5.2 and Matplotlib 1.5.1 on Linux.

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

It would not cycle if a `color` were specified. However, this does not match `plot`, which does not advance the cycle only if _all_ properties in the cycle are specified. Notably, this means if your property cycle was for line style, specifying a `color` would ignore the cycle in `errorbar`, but not in `plot`.

Fixes #7074.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py
Comment (WeatherGod) [inline lib/matplotlib/axes/_axes.py:3281]: Is this where the property cycler gets implicitly advanced?

**Comment by jklymak** [comment]
Is it the case that `errorbar` was wrong or that the behaviour for plot changed, but it didn't get changed in `errorbar`? I suspect the latter, in which case this may still get out of sync.  Does it make sense to have some sort of helper for this or add the functionality into `_get_lines._getdefaults` so this doesn't get out of sync in the future?

File: lib/matplotlib/axes/_axes.py
Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:3281]: Yes, I don't really like the private API of a private member, but that's how it is right now.

**Comment by QuLogic** [comment]
I think `_get_lines._getdefaults` is that helper.

My preferred method would be to use something higher, like  `_get_lines()` directly or `_get_lines._makeline()`, but we need the property dict to copy to the error bars. I don't think we have an easy way to copy from the `Line2D` to the error bar line. Though maybe since this is private, I can just make it return both the line and its properties?

**Comment by jklymak** [comment]
I'm not blocking on that comment - just seems if we want the two to have the same behaviour they should have the same method of getting the behaviour...

**Comment by QuLogic** [comment]
Well, it is, sort of. `self._get_lines` is the object that implements the behaviour. The `plot` call tree is `plot` -> `self._get_lines()` -> `self._get_lines._plot_args` -> `self._get_lines._makeline` -> (`self._get_lines._getdefaults` for kwargs and creates `Line2D`). Previously, errorbar used the `self._get_lines.prop_cycle` and only looked for colour; in this PR it uses `_getdefaults` like `plot` ultimately does.

I experimented a little bit with calling one of those higher methods and then `.update_from(data_line)` + overriding any properties after, but this is complicated by the kwarg inheritance in errorbar. It's currently:

* `_getdefaults` + `fmt_style_kwargs` -> `base_style`
  * \+ `kwargs` = `plot_line_style` -> the main line
  * \+ e* kwargs - markers/lines = `eb_lines_style` -> error bars
  * \+ cap settings - markers/lines = `eb_cap_style` -> error caps

However, the main line is `Line2D` while the error bars are `LineCollection`, so they have different properties and `bars.update_from(data_line)` doesn't work.

I think at best, we could have `_makeline` return both the line and its kwargs, which could then be used as `base_style`, or maybe `_plot_args`.

**Comment by QuLogic** [comment]
The closest I could do is call `self._get_lines._plot_args`, as calling `self.plot` or `self._get_lines` directly would call `_process_unit_info` again:
https://github.com/matplotlib/matplotlib/blob/bb9d11dd1a14ab918eb02c3bb80476d9aec8f153/lib/matplotlib/axes/_base.py#L210
or process `data`
https://github.com/matplotlib/matplotlib/blob/bb9d11dd1a14ab918eb02c3bb80476d9aec8f153/lib/matplotlib/axes/_base.py#L220-L263
which are not done in `errorbar`.

File: lib/matplotlib/axes/_axes.py
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py:3341]: Not quite following why all this popping needs to be done twice...

**Comment by QuLogic** [comment]
With #8031 in, I'd like to apply this to the 3D version as well before merging.

File: lib/matplotlib/axes/_axes.py
Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:3341]: What do you mean?

Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:3341]: Ah, I see, `base_style` is no longer used to create the line, so we pop everything before creating the derivative styles.

**Comment by timhoffm** [comment]
This seems rather complete (modulo some restructuring of the clipping). IMHO it would be easier for reviewing to merge this now and do the 3D stuff in a separate PR.

**Comment by QuLogic** [comment]
Sure, I ran into other issues just doing that directly, so I'll mark this Ready for Review now, and do 3D in a followup PR.

**Comment by timhoffm** [comment]
Sorry, I meant popping, not clipping.

**Comment by QuLogic** [comment]
Ah, okay, the last commit should clear that up, I hope.

**Review by timhoffm — APPROVED** [review]

**Pull request merged**
